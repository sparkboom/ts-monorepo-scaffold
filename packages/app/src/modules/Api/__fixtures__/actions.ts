export const persistRehydrateWithToken = {
  type: 'persist/REHYDRATE',
  payload: {
    api: {
      token: {
        jwt: 'PERSISTED_TOKEN',
        expiryTimeStamp: 1571615706078,
        expiresInSeconds: 3311.836,
      },
    },
    _persist: {
      version: -1,
      rehydrated: true,
    },
  },
  key: '%%APP_NAME%%',
};

export const persistRehydrateNoToken = {
  type: 'persist/REHYDRATE',
  payload: {
    api: {
      token: {},
    },
    _persist: {
      version: -1,
      rehydrated: true,
    },
  },
  key: '%%APP_NAME%%',
};
