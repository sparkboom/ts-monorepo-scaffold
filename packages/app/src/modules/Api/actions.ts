import { Action } from 'redux';

// Types
interface ApiSetTokenAction {
  (jwt: string, expiryTimeStamp: number): AppAction<'API__SET_TOKEN', { jwt: string; expiryTimeStamp: number }>;
}
interface ApiInitAction {
  (): Action<'API__INIT'>;
}

// Actions
export const apiSetToken: ApiSetTokenAction = (jwt, expiryTimeStamp) => ({
  type: 'API__SET_TOKEN',
  payload: {
    jwt,
    expiryTimeStamp,
  },
});

export const apiInit: ApiInitAction = () => ({
  type: 'API__INIT',
});
