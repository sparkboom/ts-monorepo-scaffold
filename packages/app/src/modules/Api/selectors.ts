import { createSelector } from 'reselect';
import { ClientCmdApi, ClientQueryApi, Configuration, ClientWebApi } from '@%%APP_NAME%%/api';
import { get } from 'lodash-es';
import { queryEndpointSelector, commandEndpointSelector, webEndpointSelector } from '../config/selectors';
import { Endpoint } from '../config/types';
import { Api, Token } from './types';
import { Search } from 'history';

// Types
type ApiClient = typeof ClientCmdApi | typeof ClientQueryApi | typeof ClientWebApi;
interface CreateApi {
  (Api: ApiClient, endpoint: Endpoint | null, token: Api['token']): ClientCmdApi | ClientQueryApi | ClientWebApi | null;
}

// Helpers
const createApi: CreateApi = (Api, endpoint, token) => {
  if (!endpoint || !token || !token.jwt) {
    return null;
  }
  const apiConfig = new Configuration({
    basePath: endpoint.baseUrl,
    accessToken: token.jwt,
  });
  return new Api(apiConfig);
};

// Selectors
export const routerQuerySelector = (state: RootState): Search | null => get(state, 'router.location.search', null);
export const tokenSelector = (state: RootState): Token | null => {
  const rawToken: Token | null = get(state, 'api.token', null);
  if (!rawToken) {
    return null;
  }
  return {
    ...rawToken,
    expiryTimeStamp: Number(rawToken.expiryTimeStamp),
  };
};

// Memoized Selectors
export const querySelector = createSelector(
  routerQuerySelector,
  (routerQuery: Search | null) => {
    if (!routerQuery) {
      return {};
    }
    const kvPairs = routerQuery.substr(1).split('&');
    const queryByKey = kvPairs.reduce((acc, q) => {
      const kv = q.split('=');
      return {
        ...acc,
        [kv[0]]: kv[1],
      };
    }, {});
    return queryByKey;
  },
);

export const queryApiSelector = createSelector(
  queryEndpointSelector,
  tokenSelector,
  (queryEndpoint, token) => {
    const api = createApi(ClientQueryApi, queryEndpoint, token) as ClientQueryApi;
    return api;
  },
);

export const commandApiSelector = createSelector(
  commandEndpointSelector,
  tokenSelector,
  (commandEndpoint, token) => {
    const api = createApi(ClientCmdApi, commandEndpoint, token) as ClientCmdApi;
    return api;
  },
);

export const webApiSelector = createSelector(
  webEndpointSelector,
  webEndpoint => {
    if (!webEndpoint) {
      return null;
    }
    const apiConfig = new Configuration({
      basePath: `${webEndpoint.baseUrl}${webEndpoint.basePath}`,
    });
    return new ClientWebApi(apiConfig);
  },
);
