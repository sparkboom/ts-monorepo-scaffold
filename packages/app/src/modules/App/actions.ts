import { CheckStatus, JwtError, ConfigError, Check } from './types';

// Types
interface AppInitAction {
  (): {
    type: 'APP__INIT';
    correlationId: '%%APP_NAME%%_KERNEL';
  };
}
interface AppSetCheckStatusAction {
  (check: Check, status: CheckStatus, error: JwtError | ConfigError | null): {
    type: 'APP__SET_CHECK_STATUS';
    payload: {
      check: Check;
      status: CheckStatus;
      error: string | null;
    };
  };
}

// Actions
export const appInit: AppInitAction = () => ({
  type: 'APP__INIT',
  correlationId: '%%APP_NAME%%_KERNEL',
});

export const appSetCheckStatus: AppSetCheckStatusAction = (check, status, error = null) => ({
  type: 'APP__SET_CHECK_STATUS',
  payload: {
    check,
    status,
    error,
  },
});
