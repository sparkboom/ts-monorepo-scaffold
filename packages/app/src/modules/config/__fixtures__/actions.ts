export const fullConfiguration = {
  tasks: {
    taskUpdateFrequencyMs: 10000,
  },
  featureToggles: {},
  endpoints: {
    query: {
      baseUrl: 'http://query.example.com/',
    },
    command: {
      baseUrl: 'http://cmd.example.com/',
    },
  },
};

export const partialConfiguration = {
  tasks: {
    taskUpdateFrequencyMs: 10000,
  },
  featureToggles: {},
  endpoints: {
    query: {
      baseUrl: 'http://query.example.com/',
    },
    command: {
      baseUrl: 'http://cmd.example.com/',
    },
  },
};

export const fullFeatureToggles = {
  supportMultiSelect: true,
  supportColumnSort: true,
  supportBulkReAuth: true,
  supportBulkDelete: true,
  supportFilter: true,
  supportSearch: true,
};

export const partialFeatureToggles = {
  supportMultiSelect: false,
  supportColumnSort: false,
  supportSearch: false,
};
