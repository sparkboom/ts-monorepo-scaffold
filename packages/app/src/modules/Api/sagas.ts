import {
  call,
  delay,
  apply,
  select,
  take,
  SimpleEffect,
  TakeEffectDescriptor,
  PutEffectDescriptor,
  SelectEffectDescriptor,
  CallEffectDescriptor,
  takeLatest,
  ForkEffectDescriptor,
} from 'redux-saga/effects';
import { Action } from 'redux';
import { get } from 'lodash-es';
import { ClientCmdApi, ClientQueryApi, ClientWebApi } from '@%%APP_NAME%%/api';
import { apiSetToken } from './actions';
import { querySelector, commandApiSelector, queryApiSelector, webApiSelector } from './selectors';
import { putContext } from '../Logging/effects';
import { NOTIFICATION__XTRA_LONG_EXPIRY_MS, NOTIFICATION__LONG_EXPIRY_MS } from '../Notification/actions';
import { DetailedError } from '../../common/detailedError';
import { isToken } from './types';
import { notifySaga } from '../Notification/sagas';

// Constants
export const LOG_CTX = 'API.SAGAS';
export const MIN_QUEUE_MSECS = 10000;
export const JWT_EXPIRY_BUFFER_SECS = 300;
export const MS_IN_SECONDS = 1000;
export const OBFUSCATE_TOKEN_VISIBLE_CHARS = 8;
export const ERR__TOKEN_REFRESH_TERMINATED = 'Token refresh cycle permenantly terminated.';

// Types
interface Token {
  jwt: string;
  expiryTimeStamp: number;
  expiresInSeconds: number;
}
interface TokenFromApi {
  authToken: string;
  expiresInSeconds: string;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type GeneratorNextAny = any;
type TakeEffect = SimpleEffect<'TAKE', TakeEffectDescriptor>;
type PutEffect = SimpleEffect<'PUT', PutEffectDescriptor<any>>;
type CallEffect = SimpleEffect<'CALL', CallEffectDescriptor<any>>;
type SelectEffect = SimpleEffect<'SELECT', SelectEffectDescriptor>;
type ForkEffect = SimpleEffect<'FORK', ForkEffectDescriptor<never>>;
type ApiInitSaga = Generator<TakeEffect | PutEffect | CallEffect | SelectEffect, void, GeneratorNextAny>;
type ApiSaga = Generator<ForkEffect, void, unknown>;
type RefreshTokenSaga = Generator<SelectEffect | CallEffect | PutEffect, Token | null, GeneratorNextAny>;
type GetQueryApiSaga = Generator<SelectEffect | PutEffect, ClientQueryApi | null, GeneratorNextAny>;
type GetCommandApiSaga = Generator<SelectEffect | PutEffect, ClientCmdApi | null, GeneratorNextAny>;
type GetWebApiSaga = Generator<SelectEffect | PutEffect, ClientWebApi | null, GeneratorNextAny>;
type GetSessionTokenSaga = Generator<PutEffect, Token | null, unknown>;

// Helpers
export function createToken(jwt: string | null, expiresInSeconds: string | number, date: Date | undefined = new Date()): Token | null {
  if (!jwt) {
    return null;
  }
  const expiresInSecondsNum = Number(expiresInSeconds);
  const expiryDate = date;
  expiryDate.setSeconds(expiryDate.getSeconds() + (expiresInSecondsNum || 0));
  return {
    jwt,
    expiryTimeStamp: expiryDate.getTime(),
    expiresInSeconds: expiresInSecondsNum,
  };
}

/* we're passing optional DateTime so we can easily test the function */
/**
 * getSecondsRemaining calculates the time remaining in seconds, of the timestamp provided.
 *
 * @param expiryTimeStamp - represents the number of milliseconds since the Unix epoch. This value can represent the past, present or future.
 * @param DateTime - the Date class. This parameter promotes easier unit testing.
 * @returns total seconds difference between now and the expiryTimeStamp, a negative value represents reflects a timestamp in the past,
 * and a positive value representing a future time.
 */
export function getSecondsRemaining(expiryTimeStamp: number, DateTime = Date): number {
  const nowTimeStamp = new DateTime().getTime();
  const deltaMilliseconds = expiryTimeStamp - nowTimeStamp;
  return deltaMilliseconds / MS_IN_SECONDS;
}

// obfuscateToken is used for logging
function obfuscateToken(token: Token | null): { obfuscatedToken: Token; summary: string } | null {
  if (!token) {
    return null;
  }
  const obfuscatedToken = { ...token };
  const hiddenCharLength = Math.max(2, obfuscatedToken.jwt.length - OBFUSCATE_TOKEN_VISIBLE_CHARS);
  const visibleTokenChars = obfuscatedToken.jwt.substr(hiddenCharLength);
  obfuscatedToken.jwt = '*'.repeat(hiddenCharLength) + visibleTokenChars;
  const summary = `token=*${visibleTokenChars}, ts=${obfuscatedToken.expiryTimeStamp}, expires=${obfuscatedToken.expiresInSeconds}`;
  return {
    obfuscatedToken,
    summary,
  };
}
export const refreshTokenDelayMsFromExpiry = (expiresInSeconds: number): number =>
  Math.max(MIN_QUEUE_MSECS, (expiresInSeconds - JWT_EXPIRY_BUFFER_SECS) * MS_IN_SECONDS);

// Sagas
export function* refreshTokenSaga(action: Action<string>): RefreshTokenSaga {
  const put = putContext(action, LOG_CTX);
  try {
    yield put.logDebug(`Token refresh start`);

    // 1. Get Cmd API
    const cmdApi = (yield select(commandApiSelector)) as ClientCmdApi | null;
    if (!cmdApi) {
      throw new DetailedError('Token Refresh Failed. Command API unavailable', { cmdApi });
    }

    // 2. Refresh Token
    // eslint-disable-next-line @typescript-eslint/unbound-method
    const rawToken: TokenFromApi | null = yield apply(cmdApi, cmdApi.authenticationRefreshToken, []);
    if (!rawToken) {
      throw new DetailedError('Token Refresh Failed. No valid token received from API refresh call.', { rawToken });
    }

    // 3. Transform Token
    const newToken: Token | null = yield call(createToken, rawToken.authToken, rawToken.expiresInSeconds);
    if (!newToken || !newToken.jwt) {
      throw new DetailedError('Token Refresh Failed. Token is possibly invalid.', { rawToken, newToken });
    }

    // 4. Store New Token
    yield put(apiSetToken(newToken.jwt, newToken.expiryTimeStamp));
    const obfuscatedToken = obfuscateToken(newToken);
    yield put.logInfo(`API Token successfully updated. ${obfuscatedToken && obfuscatedToken.summary}`, {
      token: obfuscatedToken && obfuscatedToken.obfuscatedToken,
    });
    if (sessionStorage) {
      sessionStorage.setItem('jwt', newToken.jwt);
      sessionStorage.setItem('expires', `${newToken.expiryTimeStamp}`);
    }

    return newToken;
  } catch (error) {
    yield put.errorWarn(error);
    yield call(
      notifySaga,
      action,
      'common.notifications.sessionRefreshFailed.heading',
      'common.notifications.sessionRefreshFailed.details',
      'ERROR',
      NOTIFICATION__XTRA_LONG_EXPIRY_MS,
    );
  }

  return null;
}

/**
 * getSessionTokenSaga obtains the available tokens from session storage, one from V1 FE app, the other from V2 FE app
 * It will compare the two and use the one that expires last, thus assuming it is the most recently retrieved token.
 *
 * When V1 FE has been retired, then we can remove this complication, and simply use 'yield select(tokenSelector)'
 *
 * @param action - triggering action for the saga
 * @param currentV2Token - we already have this value, so we'll pass it instead of retrieving it
 * @param global - typically represents the global window object. Providing the value as an argument allows this generator to be tested
 */
export function* getSessionTokenSaga(action: Action<string>, currentV2Token: Token | null, global: Window & typeof globalThis = window): GetSessionTokenSaga {
  const put = putContext(action, LOG_CTX);
  let preferredToken: Token | null = null;

  // 1. Construct V2 Token via redux-persist
  if (currentV2Token) {
    // eslint-disable-next-line no-param-reassign
    currentV2Token.expiresInSeconds = getSecondsRemaining(currentV2Token.expiryTimeStamp);
  }

  // 2. Get V1 Token via session storage if available and valid
  let currentV1Token: Token | null = null;
  const { sessionStorage } = global;
  if (sessionStorage) {
    const jwtV1 = sessionStorage.getItem('jwt');
    const expiresV1 = Number(sessionStorage.getItem('expires'));
    if (jwtV1 && expiresV1 && !Number.isNaN(expiresV1)) {
      currentV1Token = {
        jwt: jwtV1,
        expiryTimeStamp: expiresV1,
        expiresInSeconds: getSecondsRemaining(expiresV1),
      };
    }
  }

  // 3. Choose preferred token with the latter expiry
  if (!currentV1Token && currentV2Token) {
    preferredToken = currentV2Token;
  } else if (!currentV2Token && currentV1Token) {
    preferredToken = currentV1Token;
  } else if (currentV2Token && currentV1Token) {
    preferredToken = currentV1Token.expiryTimeStamp > currentV2Token.expiryTimeStamp ? currentV1Token : currentV2Token;
  } else {
    return null;
  }

  // 4. Ensure both V1 and V2 formats are the preferred token
  yield put(apiSetToken(preferredToken.jwt, preferredToken.expiryTimeStamp));
  if (sessionStorage) {
    sessionStorage.setItem('jwt', preferredToken.jwt);
    sessionStorage.setItem('expires', `${preferredToken.expiryTimeStamp}`);
  }

  return preferredToken;
}

export function* apiInitSaga(action: Action<string>, global: Window & typeof globalThis = window): ApiInitSaga {
  const put = putContext(action, LOG_CTX);
  try {
    // 1. Ensure redux-persist has rehydrated
    const persistState = yield take('persist/REHYDRATE');

    // 2. Set JWT app check to in-progress
    yield put.status('IN_PROGRESS');

    // 3. Choose token
    const searchQueries: Record<string, string> = yield select(querySelector);
    const queryToken = searchQueries ? createToken(searchQueries.jwt, searchQueries.expiresInSeconds) : null;
    let token: Token | null = null;

    debugger;

    if (isToken(queryToken)) {
      // 3a. Load token from query parameters
      // This token takes precedence, it happens on the initial redirect from the login endpoint

      // Store token in session storage in v1 & v2 format
      yield put(apiSetToken(queryToken.jwt, queryToken.expiryTimeStamp));
      if (sessionStorage) {
        sessionStorage.setItem('jwt', queryToken.jwt);
        sessionStorage.setItem('expires', `${queryToken.expiryTimeStamp}`);
      }

      // Remove query parameters, without pushing an entry in the browser navigation stack.
      const { history, document, location } = global;
      history.replaceState({}, document.title, location.origin + location.pathname);

      token = queryToken;
    } else {
      // 3b. Get most recent token from session storage - this could be v1 or v2 format
      // This happens when the user navigates using the back button, refreshes the frame, or redirects from the v1 app.
      const v2Token: Token | null = get(persistState, 'payload.api.token', null);
      token = yield* getSessionTokenSaga(action, v2Token);
    }

    // If there is no token in session storage, we cannot continue.
    if (!token) {
      throw new DetailedError('Session failed. No token found in querystring or session storage.');
    }

    // 4. Mark Token Check as success
    yield put.success();

    // 5. Token Refresh Cycle
    while (token !== null) {
      const refreshDelayMs = refreshTokenDelayMsFromExpiry(token.expiresInSeconds);
      yield put.logDebug(`Token refresh cycle started. Time until next refresh ${refreshDelayMs}ms`, {
        expiryTimeStamp: token.expiryTimeStamp,
        expiresInSeconds: token.expiresInSeconds,
      });

      yield delay(refreshDelayMs);
      token = (yield call(refreshTokenSaga, action)) as Token | null;
    }

    yield put.error(ERR__TOKEN_REFRESH_TERMINATED);
    yield call(
      notifySaga,
      action,
      'common.notifications.sessionTerminated.heading',
      'common.notifications.sessionTerminated.details',
      'ERROR',
      NOTIFICATION__LONG_EXPIRY_MS,
    );
  } catch (err) {
    yield put.error(err);
    yield call(
      notifySaga,
      action,
      'common.notifications.sessionFailed.heading',
      'common.notifications.sessionFailed.details',
      'ERROR',
      NOTIFICATION__XTRA_LONG_EXPIRY_MS,
    );
  } finally {
    yield put.fail();
  }
}

export function* getQueryApi(action: Action<any>): GetQueryApiSaga {
  const put = putContext(action, LOG_CTX);
  const queryApi: ClientQueryApi | null = yield select(queryApiSelector);

  if (!queryApi) {
    yield put.errorWarn('Query API could not be retrieved');
    return null;
  }

  return queryApi;
}

export function* getCommandApi(action: Action<any>): GetCommandApiSaga {
  const put = putContext(action, LOG_CTX);
  const commandApi: ClientCmdApi | null = yield select(commandApiSelector);

  if (!commandApi) {
    yield put.errorWarn('Command API could not be retrieved');
    return null;
  }

  return commandApi;
}

export function* getWebApi(action: Action<any>): GetWebApiSaga {
  const put = putContext(action, LOG_CTX);
  const webApi: ClientWebApi | null = yield select(webApiSelector);
  if (!webApi) {
    yield put.errorWarn('Web API could not be created');
    return null;
  }
  return webApi;
}

export default function* apiSaga(): ApiSaga {
  yield takeLatest('API__INIT', apiInitSaga);
}
