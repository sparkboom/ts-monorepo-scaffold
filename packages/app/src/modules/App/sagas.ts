import { all, takeLatest, SimpleEffect, PutEffectDescriptor, ForkEffectDescriptor, select } from 'redux-saga/effects';
import { Action } from 'redux';
import { CombinatorEffect } from '@redux-saga/types';
import { appSetCheckStatus } from './actions';
import { initTasks } from '../Tasks/actions';
import { loadContent } from '../Content/actions';
import { loadConfig, loadFeatureToggles } from '../config/actions';
import './types.ts';
import { isAppReadySelector } from './selectors';
import { apiInit } from '../Api/actions';
import { putContext } from '../Logging/effects';

// Constants
const LOG_CTX = 'APP.SAGAS';

// Types
type PutEffects = SimpleEffect<'PUT', PutEffectDescriptor<any>>;
type AllEffects = CombinatorEffect<'ALL', PutEffects>;
type AppInitSaga = Generator<AllEffects | PutEffects, void, void>;
type ForkEffect = SimpleEffect<'FORK', ForkEffectDescriptor<any>>;
type AppSaga = Generator<ForkEffect, void, void>;

// Sagas
export function* appInitSaga(action: Action<'APP__INIT'>): AppInitSaga {
  const put = putContext(action, LOG_CTX);

  yield all([
    put(appSetCheckStatus('config', 'APP_CHECK__IN_PROGRESS')), // TODO: move this to reducers
    put(loadConfig()),
    put(loadContent()),
    put(loadFeatureToggles()),
    put(apiInit()),
  ]);
}

export function* appLoadSaga(action: Action<'APP__SET_CHECK_STATUS'>): any {
  const put = putContext(action, LOG_CTX);

  const isAppReady = yield select(isAppReadySelector);
  if (isAppReady) {
    yield put(initTasks());
  }
}

function* appSaga(): AppSaga {
  yield takeLatest('APP__INIT', appInitSaga);
  yield takeLatest(['API__INIT_SUCCESS', 'CONFIG__LOAD_SUCCESS'], appLoadSaga);
}

export default appSaga;
