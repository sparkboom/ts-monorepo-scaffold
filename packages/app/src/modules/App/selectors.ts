import { createSelector } from 'reselect';
import { isString } from 'lodash-es';
import { allTasksHashSelector } from '../Tasks/selectors';

// Constants
const APP_INIT_STATUSES = ['APP_CHECK__INIT', 'APP_CHECK__IN_PROGRESS'];

// Helpers
const isStillInitializing = (check: string | string[]) => isString(check) && APP_INIT_STATUSES.includes(check);

// Selectors
export const jwtCheckStatusSelector = (state: RootState) => state.app.checks.jwt.status;
export const configCheckStatusSelector = (state: RootState) => state.app.checks.config.status;

export const isAppInitializingSelector = createSelector(
  jwtCheckStatusSelector,
  configCheckStatusSelector,
  (jwtCheck, configCheck) => isStillInitializing(jwtCheck) || isStillInitializing(configCheck),
);

export const isAppReadySelector = createSelector(
  jwtCheckStatusSelector,
  configCheckStatusSelector,
  (jwtCheck, configCheck) => jwtCheck === 'APP_CHECK__SUCCESS' && configCheck === 'APP_CHECK__SUCCESS',
);

export const areTasksReadySelector = createSelector(
  isAppReadySelector,
  allTasksHashSelector,
  (isAppReady, allTasksHash) => isAppReady && allTasksHash !== null,
);
