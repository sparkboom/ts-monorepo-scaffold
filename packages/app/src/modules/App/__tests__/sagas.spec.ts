import { expectSaga } from 'redux-saga-test-plan';
import { select } from 'redux-saga/effects';
import * as actions from '../actions';
import { isAppReadySelector } from '../selectors';
import { initTasks } from '../../Tasks/actions';
import { loadContent } from '../../Content/actions';
import { loadConfig, loadFeatureToggles } from '../../config/actions';
import * as sagas from '../sagas';
import { apiInit } from '../../Api/actions';

describe('App > Sagas', () => {
  let FAKE_ACTION: ReturnType<typeof actions.appInit>;

  beforeEach(() => {
    FAKE_ACTION = actions.appInit();
  });

  describe('appInitSaga', () => {
    it('should initiated configuration check to `in progress`', () => {
      // Arrange
      // Act
      // Assert
      return expectSaga(sagas.appInitSaga, FAKE_ACTION)
        .put.like({ action: actions.appSetCheckStatus('config', 'APP_CHECK__IN_PROGRESS', null) })
        .run();
    });

    it('should dispatch load config action', () => {
      // Arrange
      // Act
      // Assert
      return expectSaga(sagas.appInitSaga, FAKE_ACTION)
        .put.like({ action: loadConfig() })
        .run();
    });

    it('should dispatch load content action', () => {
      // Arrange
      // Act
      // Assert
      return expectSaga(sagas.appInitSaga, FAKE_ACTION)
        .put.like({ action: loadContent() })
        .run();
    });

    it('should dispatch load feature toggles action', () => {
      // Arrange
      // Act
      // Assert
      return expectSaga(sagas.appInitSaga, FAKE_ACTION)
        .put.like({ action: loadFeatureToggles() })
        .run();
    });

    it('should dispatch api initialization action', () => {
      // Arrange
      // Act
      // Assert
      return expectSaga(sagas.appInitSaga, FAKE_ACTION)
        .put.like({ action: apiInit() })
        .run();
    });

    it('should match snapshot', async () => {
      // Arrange
      // Act
      // Assert
      const result = await expectSaga(sagas.appInitSaga, FAKE_ACTION)
        .put.like({ action: apiInit() })
        .run();

      expect(result.toJSON()).toMatchSnapshot();
    });
  });

  describe('appLoadSaga', () => {
    it('should not dispatch init tasks action if app is not ready', () => {
      // Arrange
      const FAKE_IS_APP_READY = false;

      // Act
      // Assert
      return expectSaga(sagas.appLoadSaga, FAKE_ACTION)
        .provide([[select(isAppReadySelector), FAKE_IS_APP_READY]])
        .not.put.like({ action: initTasks() })
        .run();
    });

    it('should dispatch init tasks action if app is ready', () => {
      // Arrange
      const FAKE_IS_APP_READY = true;

      // Act
      // Assert
      return expectSaga(sagas.appLoadSaga, FAKE_ACTION)
        .provide([[select(isAppReadySelector), FAKE_IS_APP_READY]])
        .put.like({ action: initTasks() })
        .run();
    });

    it('should match snapshot if app is ready', async () => {
      // Arrange
      const FAKE_IS_APP_READY = true;

      // Act
      // Assert
      const result = await expectSaga(sagas.appLoadSaga, FAKE_ACTION)
        .provide([[select(isAppReadySelector), FAKE_IS_APP_READY]])
        .put.like({ action: initTasks() })
        .run();

      expect(result.toJSON()).toMatchSnapshot();
    });
  });
});
