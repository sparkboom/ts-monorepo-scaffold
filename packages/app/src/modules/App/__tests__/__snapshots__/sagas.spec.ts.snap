// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App > Sagas appInitSaga should match snapshot 1`] = `
Object {
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "correlationId": "%%APP_NAME%%_KERNEL",
          "payload": Object {
            "check": "config",
            "error": null,
            "status": "APP_CHECK__IN_PROGRESS",
          },
          "type": "APP__SET_CHECK_STATUS",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "correlationId": "%%APP_NAME%%_KERNEL",
          "type": "CONFIG__LOAD",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "correlationId": "%%APP_NAME%%_KERNEL",
          "type": "CONTENT__LOAD",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "correlationId": "%%APP_NAME%%_KERNEL",
          "type": "CONFIG__FEATURE_TOGGLES_LOAD",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
}
`;

exports[`App > Sagas appLoadSaga should match snapshot if app is ready 1`] = `
Object {
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
  ],
}
`;
